csharp/src/Ice/SlicStream.cs

```diff
@@ -270,13 +257,115 @@ async Task SendFrameAsync(int frameSize, bool fin, IList<ArraySegment<byte>> buf
         }
 
         internal SlicStream(long streamId, SlicTransceiver transceiver)
-            : base(streamId, transceiver) => _transceiver = transceiver;
+            : base(streamId, transceiver)
+        {
+            _transceiver = transceiver;
+
+            if (IsIncoming && !IsControl)
+            {
+                // Increment the counters to ensure the peer doesn't open more streams than allowed.
+                Debug.Assert(IsIncoming);
+                if (IsBidirectional)
+                {
+                    if (_transceiver.BidirectionalStreamCount == _transceiver.Options.MaxBidirectionalStreams)
+                    {
+                        throw new InvalidDataException(
+                            $"maximum bidirectional stream count {_transceiver.Options.MaxBidirectionalStreams} reached");
+                    }
+                    Interlocked.Increment(ref _transceiver.BidirectionalStreamCount);
```

Comment:
couldn't this result in a race condition, between the check and the increment?

Reply:
Good point, I think it's fine because the constructor can't be called concurrently on a given connection. Added the following to the comment:
```
                // Increment the counter to ensure the peer doesn't open more streams than allowed. This can't be
                // called concurrently for a given connection so if it's fine to check the counter and increment
                // it after.
```